txscript/sigcache.go

```diff
@@ -141,3 +146,59 @@ func shortTxHash(msg *wire.MsgTx, key [shortTxHashKeySize]byte) uint64 {
 	txHash := msg.TxHash()
 	return siphash.Hash(k0, k1, txHash[:])
 }
+
+// EvictEntries removes all entries from the SigCache that correspond to the
+// transactions in the given block.  The block that is passed must be
+// ProactiveEvictionDepth blocks deep, which is the depth at which the
+// signatures for the transactions within the block are nearly guaranteed to no
+// longer be useful.
+//
+// EvictEntries wraps the unexported evictEntries method, which is run from a
+// goroutine.  evictEntries is only invoked if validSigs is not empty.  This
+// avoids starting a new goroutine when there is nothing to evict, such as when
+// syncing is ongoing.
+func (s *SigCache) EvictEntries(block *wire.MsgBlock) {
+	s.RLock()
+	if len(s.validSigs) == 0 {
+		return
```

Comment:
This will result in a deadlock when there are no valid signatures here because the mutex isn't being released.
```suggestion
		s.RUnlock()
		return
```

Reply:
Good catch, thanks.  Fixed.